File: services/corporate-events/src/domain/models/stock_investment.py
from dataclasses import dataclass, field
from decimal import Decimal
import datetime as dt


@dataclass
class StockInvestment:
    subject: str
    id: str
    date: dt.date
    broker: str
    ticker: str
    amount: Decimal
    price: Decimal
    costs: Decimal = field(default_factory=lambda: Decimal(0))
    alias_ticker: str = ""
    external_system: str = ""
    operation: str = "BUY"
    type: str = "STOCK"

    def __post_init__(self):
        if not isinstance(self.amount, Decimal):
            self.amount = Decimal(self.amount).quantize(Decimal("0.01"))
        if not isinstance(self.price, Decimal):
            self.price = Decimal(self.price).quantize(Decimal("0.01"))
        if not isinstance(self.costs, Decimal):
            self.costs = Decimal(self.costs).quantize(Decimal("0.01"))

    @property
    def current_ticker_name(self):
        return self.alias_ticker or self.ticker

File: mobile/lib/services/investment/model/stock_investment.dart
import 'package:goatfolio/services/investment/model/investment.dart';
import 'package:intl/intl.dart';

class StockInvestment extends Investment {
  String operation;
  String? broker;
  double costs;
  String ticker;
  int amount;
  double price;
  String? aliasTicker;

  StockInvestment({
    required this.ticker,
    required this.amount,
    required this.price,
    required String type,
    required this.operation,
    required DateTime date,
    String? id,
    String? subject,
    required this.broker,
    required this.costs,
    this.aliasTicker
  }): super(id: id, subject: subject, date: date, type: type);

  void copy(StockInvestment inv) {
    this.ticker = inv.ticker;
    this.amount = inv.amount;
    this.price = inv.price;
    this.type = inv.type;
    this.operation = inv.operation;
    this.date = inv.date;
    this.id = inv.id;
    this.subject = inv.subject;
    this.broker = inv.broker;
    this.costs = inv.costs;
  }

  factory StockInvestment.fromJson(Map<String, dynamic> json) => StockInvestment(
      operation: json['operation'],
      type: json['type'],
      date: DateTime.parse('${json['date']}'),
      id: json['id'],
      subject: json['subject'],
      broker: json['broker'],
      ticker: json['ticker'],
      amount: json['amount'].toInt(),
      price: json['price'],
      costs: json['costs'],
      aliasTicker: json['alias_ticker']
  );

  Map<String, dynamic> toJson() => {
        'type': type,
        'operation': operation,
        'date': DateFormat("yyyyMMdd").format(date),
        'broker': broker,
        'id': id,
        'ticker': ticker,
        'price': price,
        'amount': amount,
        'costs': costs,
        'alias_ticker': aliasTicker
      };
}
